File: Easy/1796_Second_Largest_Digit_in_a_String/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int secondHighest(string s) {
        int fastLarge = -1;
        int secendLarge = -1;

        for (int i = 0; i < s.size(); i++) {
            if (isdigit(s[i])) {
                int num = s[i] - '0';

                if (num > fastLarge) {
                    secendLarge = fastLarge;
                    fastLarge = num;
                } else if (num < fastLarge && num > secendLarge) {
                    secendLarge = num;
                }
            }
        }

        return secendLarge;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    cout << solution.secondHighest("dfa12321afd") << endl;

    // test cases 2
    cout << solution.secondHighest("abc1111") << endl;

    return 0;
}

File: Easy/1796_Second_Largest_Digit_in_a_String/Solution.java


public class Solution {
    public int secondHighest(String s) {

        int fastLarge = -1;
        int secendLarge = -1;

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);

            if (Character.isDigit(ch)) {
                int val = ch - '0';

                if (val > fastLarge) {
                    secendLarge = fastLarge;
                    fastLarge = val;
                } else if (val < fastLarge && val > secendLarge) {
                    secendLarge = val;
                }
            }
        }
        
        return secendLarge;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String s1 = "dfa12321afd";
        System.out.println(solution.secondHighest(s1)); // Output: 2
        
        // test cases 2
        String s2 = "abc1111";
        System.out.println(solution.secondHighest(s2)); // Output: -1
        
    }
}